sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -361,6 +361,18 @@ object KeyGroupedPartitioning {
       partitionValues: Seq[InternalRow]): KeyGroupedPartitioning = {
     KeyGroupedPartitioning(expressions, partitionValues.size, Some(partitionValues))
   }
+
+  def supportsExpressions(expressions: Seq[Expression]): Boolean = {
```

Comment:
@sunchao @cloud-fan, I went back and forth on where to add this validation. I decided to add it here as it is a current limitation of internal Catalyst `KeyGroupedPartitioning`. It is fine if a data source reports a partitioning with multi-arg transforms, we just can't benefit from it right now.

Let me know what you think. I also added a test to `KeyGroupedPartitioningSuite`.

Reply:
I guess, this code is supposed to make sure there is only one child, but it was skipped in my execution as `sorted` was empty, though I don't understand what is this requirement for:
https://github.com/apache/spark/pull/35657/files#diff-715d0c2d59a4ddb8a4b5952c1b05be9f035b6d9b0d9670c70b58989dc722b252R86
Making `sorted` empty let me pass this check, though I need it for the join optimization. But I have managed to bring this property back on the physical level using `org.apache.spark.sql.connector.read.SupportsReportOrdering` and execute the `SortMergeJoin` without the exchanges and sorts.